File: xml/src/test/java/akka/stream/alpakka/xml/javadsl/XmlWritingTest.java
/*
 * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
 */

package akka.stream.alpakka.xml.javadsl;

import akka.actor.ActorSystem;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import akka.stream.alpakka.xml.*;
import akka.stream.javadsl.Flow;
import akka.stream.javadsl.Keep;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import akka.testkit.JavaTestKit;
import akka.util.ByteString;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import static org.junit.Assert.assertEquals;

public class XmlWritingTest {
  private static ActorSystem system;
  private static Materializer materializer;

  @Test
  public void xmlWriter() throws InterruptedException, ExecutionException, TimeoutException {

    // #writer
    final Sink<ParseEvent, CompletionStage<String>> write = Flow.of(ParseEvent.class)
      .via(XmlWriting.writer())
      .map((ByteString bs) -> bs.utf8String())
      .toMat(Sink.fold("", (acc, el) -> acc + el), Keep.right());
    // #writer

    // #writer-usage
    final String doc = "<?xml version='1.0' encoding='UTF-8'?><doc><elem>elem1</elem><elem>elem2</elem></doc>";
    final List<ParseEvent> docList= new ArrayList<ParseEvent>();
    docList.add(StartDocument.getInstance());
    docList.add(StartElement.create("doc", Collections.emptyMap()));
    docList.add(StartElement.create("elem", Collections.emptyMap()));
    docList.add(Characters.create("elem1"));
    docList.add(EndElement.create("elem"));
    docList.add(StartElement.create("elem", Collections.emptyMap()));
    docList.add(Characters.create("elem2"));
    docList.add(EndElement.create("elem"));
    docList.add(EndElement.create("doc"));
    docList.add(EndDocument.getInstance());


    final CompletionStage<String> resultStage = Source.from(docList).runWith(write, materializer);
    // #writer-usage

    resultStage.thenAccept((str) -> {
      assertEquals(doc,str);
    }).toCompletableFuture().get(5, TimeUnit.SECONDS);
  }

  @BeforeClass
  public static void setup() throws Exception {
    system = ActorSystem.create();
    materializer = ActorMaterializer.create(system);
  }

  @AfterClass
  public static void teardown() throws Exception {
    JavaTestKit.shutdownActorSystem(system);
  }
}

File: xml/src/test/scala/akka/stream/alpakka/xml/scaladsl/XmlWritingTest.scala
/*
 * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
 */

package akka.stream.alpakka.xml.scaladsl

import akka.actor.ActorSystem
import akka.stream.ActorMaterializer
import akka.stream.alpakka.xml._
import akka.stream.scaladsl.{Flow, Keep, Sink, Source}
import org.scalatest.concurrent.PatienceConfiguration.Timeout
import org.scalatest.concurrent.ScalaFutures
import org.scalatest.{BeforeAndAfterAll, Matchers, WordSpec}

import scala.concurrent.Future
import scala.concurrent.duration._

class XmlWritingTest extends WordSpec with Matchers with BeforeAndAfterAll with ScalaFutures {
  implicit val system = ActorSystem("Test")
  implicit val mat = ActorMaterializer()

  // #writer
  val writer: Sink[ParseEvent, Future[String]] = Flow[ParseEvent]
    .via(XmlWriting.writer)
    .map[String](_.utf8String)
    .toMat(Sink.fold[String, String]("")((t, u) => t + u))(Keep.right)
  // #writer

  "XML Writer" must {

    "properly write simple XML" in {
      // #writer-usage
      val listEl: List[ParseEvent] = List(
        StartDocument,
        StartElement("doc", Map.empty),
        StartElement("elem", Map.empty),
        Characters("elem1"),
        EndElement("elem"),
        StartElement("elem", Map.empty),
        Characters("elem2"),
        EndElement("elem"),
        EndElement("doc"),
        EndDocument
      )

      val doc = "<?xml version='1.0' encoding='UTF-8'?><doc><elem>elem1</elem><elem>elem2</elem></doc>"
      val resultFuture: Future[String] = Source.fromIterator[ParseEvent](() => listEl.iterator).runWith(writer)
      // #writer-usage

      resultFuture.futureValue(Timeout(20.seconds)) should ===(doc)
    }

    "properly process a comment" in {
      val doc = "<?xml version='1.0' encoding='UTF-8'?><doc><!--comment--></doc>"
      val listEl = List(
        StartDocument,
        StartElement("doc", Map.empty),
        Comment("comment"),
        EndElement("doc"),
        EndDocument
      )

      val resultFuture: Future[String] = Source.fromIterator[ParseEvent](() => listEl.iterator).runWith(writer)

      resultFuture.futureValue(Timeout(20.seconds)) should ===(doc)
    }

    "properly process parse instructions" in {
      val doc = """<?xml version='1.0' encoding='UTF-8'?><?target content?><doc/>"""
      val listEl = List(
        StartDocument,
        ProcessingInstruction(Some("target"), Some("content")),
        StartElement("doc", Map.empty),
        EndElement("doc"),
        EndDocument
      )

      val resultFuture: Future[String] = Source.fromIterator[ParseEvent](() => listEl.iterator).runWith(writer)

      resultFuture.futureValue(Timeout(20.seconds)) should ===(doc)

    }

    "properly process attributes" in {
      val doc =
        """<?xml version='1.0' encoding='UTF-8'?><doc good="yes"><elem nice="yes" very="true">elem1</elem></doc>"""
      val listEl = List(
        StartDocument,
        StartElement("doc", Map("good" -> "yes")),
        StartElement("elem", Map("nice" -> "yes", "very" -> "true")),
        Characters("elem1"),
        EndElement("elem"),
        EndElement("doc"),
        EndDocument
      )

      val resultFuture: Future[String] = Source.fromIterator[ParseEvent](() => listEl.iterator).runWith(writer)

      resultFuture.futureValue(Timeout(3.seconds)) should ===(doc)
    }

    "properly process CData blocks" in {
      val doc = """<?xml version='1.0' encoding='UTF-8'?><doc><![CDATA[<not>even</valid>]]></doc>"""

      val listEl = List(
        StartDocument,
        StartElement("doc", Map.empty),
        CData("<not>even</valid>"),
        EndElement("doc"),
        EndDocument
      )
      val resultFuture: Future[String] = Source.fromIterator[ParseEvent](() => listEl.iterator).runWith(writer)

      resultFuture.futureValue(Timeout(3.seconds)) should ===(doc)
    }

  }

  override protected def afterAll(): Unit = system.terminate()
}
